### PR TITLE
Extract propTypes detection code

### PR DESCRIPTION
**Pull request by alexzherdev**
~~This still fails a bunch of tests, but wanted to put it up for preliminary high-level review early.~~

~~The reason tests are failing is because I extracted similar code from `prop-types` and `no-unused-prop-types` but those two have diverged and I need to reconcile two implementations into a single helper.~~

Next steps:

- [x] fix failing tests (this PR)
- [x] look at helper coverage when separately running tests for both rules and see if any one of them is missing test cases that they now support by virtue of using a unified implementation _(PR in my fork: alexzherdev/eslint-plugin-react#1)_
- [x] currently only **declared** propTypes detection is extracted. Could also extract **used** propTypes detection from `prop-types` and `no-unused-prop-types` ? _(PR in my fork: alexzherdev/eslint-plugin-react#2)_
- [x] `require-default-props` and `default-props-match-prop-types` also have some common functionality for requiredness and defaultProps detection - extract those? That would be a prerequisite to those rules using the common helper. _(PRs in my fork: alexzherdev/eslint-plugin-react#3, alexzherdev/eslint-plugin-react#4)_

### CODE REVIEW

File: lib/rules/no-unused-prop-types.js [hunk 1]
```diff
@@ -893,7 +454,7 @@ module.exports = {
         return;
       }
 
-      (props || []).forEach(prop => {
+      Object.values(props || {}).forEach(prop => {
```
Comment (alexzherdev) [inline lib/rules/no-unused-prop-types.js]: `declaredPropTypes` used to be an array here, but an object in `prop-types`. Common implementation uses an object (keyed by prop name) for easier lookups

File: lib/rules/no-unused-prop-types.js [hunk 2]
```diff
@@ -980,28 +509,31 @@ module.exports = {
     function handleFunctionLikeExpressions(node) {
       handleSetStateUpdater(node);
       markDestructuredFunctionArgumentsAsUsed(node);
-      markAnnotatedFunctionArgumentsAsDeclared(node);
+    }
+
+    function handleCustomValidators(component) {
+      const propTypes = component.declaredPropTypes;
```
Comment (alexzherdev) [inline lib/rules/no-unused-prop-types.js]: This piece of code deals with used propTypes, but it was nested inside of `markPropTypesAsDeclared` (which was extracted out).

File: lib/util/propTypes.js [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+module.exports = {
+  instructions(context, components, utils) {
+    let stack = null;
+
+    // Used to track the type annotations in scope.
+    // Necessary because babel's scopes do not track type annotations.
+    const classExpressions = [];
+    const defaults = {skipShapeProps: true, customValidators: []};
+    const configuration = Object.assign({}, defaults, context.options[0] || {});
+    const customValidators = configuration.customValidators;
```
Comment (alexzherdev) [inline lib/util/propTypes.js]: These two config options come from two different rules. Put them here for now, maybe will need to consider if all rules using this functionality should take these as options:

File: lib/util/propTypes.js [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+    const skipShapeProps = configuration.skipShapeProps;
+    const sourceCode = context.getSourceCode();
+    const propWrapperFunctions = new Set(context.settings.propWrapperFunctions || []);
```
Comment (ljharb) [inline lib/util/propTypes.js]: `new Set` takes null and undefined quietly, so i'm not sure the `|| []` is needed.

File: lib/util/propTypes.js [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+
+    /**
+     * Helper for accessing the current scope in the stack.
+     * @param {string} key The name of the identifier to access. If omitted, returns the full scope.
+     * @param {ASTNode} value If provided sets the new value for the identifier.
+     * @returns {Object|ASTNode} Either the whole scope or the ASTNode associated with the given identifier.
+     */
+    function typeScope(key, value) {
+      if (arguments.length === 0) {
```
Comment (ljharb) [inline lib/util/propTypes.js]: is there another way to write this function so it doesn't use `arguments`?

File: lib/util/propTypes.js [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+        // prevent problems with recursive types.
+        seen = new Set();
+      }
+      if (seen.has(annotation)) {
+        // This must be a recursive type annotation, so just accept anything.
+        return {};
+      }
+      seen.add(annotation);
+
+      switch (annotation.type) {
```
Comment (ljharb) [inline lib/util/propTypes.js]: not a huge deal, but it'd be great to avoid a switch statement

File: lib/util/propTypes.js [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+
+module.exports = {
+  instructions(context, components, utils) {
+    let stack = null;
+
+    // Used to track the type annotations in scope.
+    // Necessary because babel's scopes do not track type annotations.
+    const classExpressions = [];
+    const defaults = {skipShapeProps: true, customValidators: []};
+    const configuration = Object.assign({}, defaults, context.options[0] || {});
```
[the remark this answers is not in the document]
Comment (alexzherdev) [reply lib/util/propTypes.js]: that's not compatible with old node versions though?

File: lib/util/propTypes.js (hunk 5 above)
Comment (alexzherdev) [reply lib/util/propTypes.js]: without changing it too much, I could check for `key` and/or `value` to be undefined I guess

File: lib/util/propTypes.js (hunk 6 above)
Comment (alexzherdev) [reply lib/util/propTypes.js]: hmm do you mean a map with handlers for each type? there's also a bunch of other huge switches in here (like in other parts of the codebase) which could also get the same treatment

File: lib/util/propTypes.js (hunk 3 above)
Comment (alexzherdev) [reply lib/util/propTypes.js]: Actually I was partly wrong, `customValidators` is present in both `prop-types` and `no-unused-prop-types`. `skipShapeProps` comes from `no-unused-prop-types` only, because it can be hard to figure out that a prop access corresponds to a shape. Not sure it's needed in `prop-types`, will take a look.

One problem right now is `skipShapeProps` is `true` by default in `no-unused-prop-types`, whereas `prop-types` never ignores shape props. Will try to fix that by always building declaredPropTypes for shapes and checking the option when validating props.

**Comment by alexzherdev** [comment]
@ljharb thanks for the review. I'm still changing things up here fixing tests, just wanted to make sure I'm not doing anything outlandish.

**Comment by alexzherdev** [comment]
screen shot 2018-08-03 at 9 07 50 pm
screen shot 2018-08-03 at 9 08 05 pm

Github bug? 🤔(same comment listed twice, one doesn't have a reply option)

File: lib/util/propTypes.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (ljharb) [reply lib/util/propTypes.js]: oh right, if we're not using babel here then we have to stick to Object.assign

File: lib/util/propTypes.js (hunk 5 above)
Comment (ljharb) [reply lib/util/propTypes.js]: an alternative is creating multiple functions rather than overloading this one?

File: lib/util/propTypes.js (hunk 6 above)
Comment (ljharb) [reply lib/util/propTypes.js]: if/else, or a map, yes

Comment (alexzherdev) [reply lib/util/propTypes.js]: Not sure how 4-5 chained else-ifs are better than a switch? anyway will try out a map

Comment (ljharb) [reply lib/util/propTypes.js]: they don't suffer from unintentional fallthrough.

**Comment by alexzherdev** [comment]
@ljharb I responded to your comments. Could you take a look at the items in the PR description to see if they make sense? If they do, should I continue modifying this PR, or would it get too big?

File: lib/rules/display-name.js [hunk 8]
```diff
@@ -116,7 +117,7 @@ module.exports = {
       );
 
       const namedFunctionExpression = (
-        (node.type === 'FunctionExpression' || node.type === 'ArrowFunctionExpression') &&
+        astUtil.isFunctionLikeExpression(node) &&
```
Comment (ljharb) [inline lib/rules/display-name.js]: I think this kind of change could be peeled off into a separate pr, and landed first?

Comment (alexzherdev) [reply lib/rules/display-name.js]: Ah, you’re right 👍

**Comment by alexzherdev** [comment]
For those other changes though all of them depend on this one to be merged first.

I already have a bunch of new tests for item 2, I can push those separately then if you’d like.

I’ll split out the function helper, and then this PR can be considered complete.

**Comment by alexzherdev** [comment]
@ljharb by itself this PR won’t fix #1908 as it doesn’t touch that particular rule.
My comment in #1908 might be misleading though. This detection doesn’t deal with defaultProps per se, but it does have the means to handle literals as propType keys. So the broader point I was trying to make is that having centralized defaultProps handling would make things like that just work. And this PR will provide the base for that further work (if we can get other collabs to sign off on it)
Sorry for the confusion, I hope this is more clear now.

**Comment by alexzherdev** [comment]
@yannickcr @lencioni @EvHaus any chance you could take a look? asking because I have a truckload of sweet refactoring elsewhere that all depends on this one (see PR description) and I really like how it turned out.

**Comment by ljharb** [comment]
I'll merge tonight if no issues are raised.
